### PR TITLE
openapi: utilize media type examples if available

### PR DESCRIPTION
**Pull request by yangguangxiadeyun**
What do think about if we support OpenAPI examples as a generated request body if it's available ?

### CODE REVIEW

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java [hunk 1]
```diff
@@ -88,4 +102,39 @@ private String generateBody() {
         }
         return "";
     }
+
+    private String extractExampleBody(Content content, String mediaTypeString) {
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java:106]: This should be part of `BodyGenerator` (which already uses examples in some cases).

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java [hunk 2]
```diff
@@ -88,4 +102,39 @@ private String generateBody() {
...
+                    content.get(mediaTypeString).getExamples().entrySet().stream()
+                            .findFirst()
+                            .get()
+                            .getValue();
+
+            if (example.getValue() != null) {
+                exampleBody = example.getValue().toString();
+            } else if (example.getExternalValue() != null
+                    && !example.getExternalValue().isEmpty()) {
+                Requestor requestor = new Requestor(HttpSender.MANUAL_REQUEST_INITIATOR);
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java:126]: Not sure about doing these extra requests. In any case this is not correct when the definition is imported using the spider.

**Comment by yangguangxiadeyun** [comment]
@thc202 please help to review again. Thank you.

**Comment by thc202** [comment]
The changelog should be updated.
The code is not properly formatted (run Gradle task `spolessApply`).

**Comment by yangguangxiadeyun** [comment]
> 
> 
> The changelog should be updated.
> The code is not properly formatted (run Gradle task `spotlessApply`).

Thanks @thc202 . I've now updated code format and change log. Please review.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java [hunk 3]
```diff
@@ -35,6 +38,7 @@
 
     private OperationModel operationModel;
     private Generators generators;
+    private static final Logger LOG = LogManager.getLogger(RequestModelConverter.class);
```
Comment (kingthorin) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java:41]: This doesn't seem to be used in the new code?

Comment (yangguangxiadeyun) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java:41]: Right. I will remove it. Thanks.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java [hunk 4]
```diff
@@ -64,8 +68,8 @@ private String generateBody() {
             Schema<?> schema;
 
             if (content.containsKey("application/json")) {
-                schema = content.get("application/json").getSchema();
-                return generators.getBodyGenerator().generate(schema);
+                MediaType mediaType = content.get("application/json");
+                return generators.getBodyGenerator().generate(mediaType);
```
[the remark this answers is not in the document]
Comment (yangguangxiadeyun) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java:72]: Right! I'll correct it.

[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java:72]: Might need to spotlessApply to get the wrapping right 😉

[the remark this answers is not in the document]
Comment (yangguangxiadeyun) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java]: Thanks. Please check again.

**Comment by yangguangxiadeyun** [comment]
@thc202 @kingthorin  please take a look again. Thank you.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 5]
```diff
@@ -339,4 +346,23 @@ private static String urlEncode(String string) {
             return "";
         }
     }
+
+    private String extractExampleBody(MediaType mediaType) {
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:350]: `static`

File: addOns/openapi/CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ### Fixed
  - Fixed var support in URLs ([Issue #6726](https://github.com/zaproxy/zaproxy/issues/6726))
 
+### Changed
+- Enhance BodyGenerator to use openapi examples if available
```
Comment (thc202) [inline addOns/openapi/CHANGELOG.md:11]:
```suggestion
- Use `application/json` media type examples when available.
```

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 7]
```diff
@@ -339,4 +346,23 @@ private static String urlEncode(String string) {
...
+                && !mediaType.getExamples().isEmpty()
+                && mediaType.getExamples().entrySet().stream().findFirst().get().getValue()
+                        != null) {
+            Example example =
+                    mediaType.getExamples().entrySet().stream().findFirst().get().getValue();
+
+            if (example.getValue() != null) {
+                exampleBody = example.getValue().toString();
+            }
+        } else if (mediaType.getExample() != null) {
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:363]: Could you add a test case for this one too?

File: addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/v3/BodyGeneratorUnitTest.java [hunk 8]
```diff
@@ -566,6 +566,22 @@ void shouldUseExample() throws IOException {
         assertEquals("{\"age\":3,\"name\":\"Fluffy\"}", request);
     }
 
+    @Test
+    void shouldUseExamples() throws IOException {
```
Comment (thc202) [inline addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/v3/BodyGeneratorUnitTest.java:570]: `shouldUseJsonMediaTypeExamples`

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 9]
```diff
@@ -339,4 +346,23 @@ private static String urlEncode(String string) {
...
+            exampleBody = mediaType.getExample().toString();
+        }
+        return exampleBody;
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:366]:
```suggestion
        return Optional.ofNullable(mediaType.getExamples())
                .map(Map::values)
                .map(Collection::stream)
                .map(stream -> stream.map(Example::getValue).filter(Objects::nonNull).findFirst())
                .orElse(Optional.ofNullable(mediaType.getExample()))
                .map(Object::toString)
                .orElse(null);
```

File: addOns/openapi/src/test/resources/org/zaproxy/zap/extension/openapi/v3/OpenApi_defn_examples.yaml [hunk 10]
```diff
@@ -22,6 +22,28 @@ paths:
...
+            schema:
+              $ref: '#/components/schemas/Pet'
+            examples:
+              fluffy-example:
+                description: a pet called Big Fluffy
+                value:
+                  age: 6
+                  name: Big Fluffy
+              fawkes-example:
+                description: a pet called Small Fawkes
```
Comment (thc202) [inline addOns/openapi/src/test/resources/org/zaproxy/zap/extension/openapi/v3/OpenApi_defn_examples.yaml:40]: Not that it matters for the code/tests but it's actually called `Small Fluffy`.

File: addOns/openapi/CHANGELOG.md (hunk 6 above)
Comment (yangguangxiadeyun) [reply addOns/openapi/CHANGELOG.md:11]: Updated

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java (hunk 7 above)
Comment (yangguangxiadeyun) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:363]: Sure

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java (hunk 9 above)
Comment (yangguangxiadeyun) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:366]: updated. Elegant code!

File: addOns/openapi/src/test/resources/org/zaproxy/zap/extension/openapi/v3/OpenApi_defn_examples.yaml (hunk 10 above)
Comment (yangguangxiadeyun) [reply addOns/openapi/src/test/resources/org/zaproxy/zap/extension/openapi/v3/OpenApi_defn_examples.yaml:40]: updated.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

Commits could have been squashed (though can be done on merge too).

**Comment by kingthorin** [comment]
Fixed up.

**Review by kingthorin — APPROVED** [review]
Thanks!

LGTM

**Comment by thc202** [comment]
Thank you both!

**Pull request merged**